File: backend/routes/properties.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from typing import Optional
from models import Property, PropertyCreate, PropertyUpdate
from auth import get_current_user_id
from database import get_db
from supabase import Client

router = APIRouter(prefix="/properties", tags=["properties"])


@router.get("/", response_model=list[Property])
def list_properties(
    search: Optional[str] = Query(None),
    bedrooms: Optional[int] = Query(None),
    max_price: Optional[float] = Query(None),
    status: Optional[str] = Query(None),
    user_id: str = Depends(get_current_user_id),
    db: Client = Depends(get_db),
):
    query = db.table("properties").select("*").eq("user_id", user_id)

    if search:
        query = query.or_(
            f"address.ilike.%{search}%,suburb.ilike.%{search}%"
        )
    if bedrooms:
        if bedrooms >= 3:
            query = query.gte("bedrooms", bedrooms)
        else:
            query = query.eq("bedrooms", bedrooms)
    if max_price:
        query = query.lte("price_per_week", max_price)
    if status and status != "":
        query = query.eq("status", status)

    result = query.order("created_at", desc=True).execute()
    return result.data


@router.post("/", response_model=Property, status_code=status.HTTP_201_CREATED)
def create_property(
    data: PropertyCreate,
    user_id: str = Depends(get_current_user_id),
    db: Client = Depends(get_db),
):
    payload = data.model_dump()
    payload["user_id"] = user_id

    result = db.table("properties").insert(payload).execute()
    if not result.data:
        raise HTTPException(status_code=500, detail="Failed to create property")
    return result.data[0]


@router.get("/{property_id}", response_model=Property)
def get_property(
    property_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Client = Depends(get_db),
):
    result = (
        db.table("properties")
        .select("*")
        .eq("id", property_id)
        .eq("user_id", user_id)
        .single()
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=404, detail="Property not found")
    return result.data


@router.put("/{property_id}", response_model=Property)
def update_property(
    property_id: str,
    data: PropertyUpdate,
    user_id: str = Depends(get_current_user_id),
    db: Client = Depends(get_db),
):
    payload = data.model_dump(exclude_none=True)
    if not payload:
        raise HTTPException(status_code=400, detail="No fields to update")

    result = (
        db.table("properties")
        .update(payload)
        .eq("id", property_id)
        .eq("user_id", user_id)
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=404, detail="Property not found")
    return result.data[0]


@router.delete("/{property_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_property(
    property_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Client = Depends(get_db),
):
    result = (
        db.table("properties")
        .delete()
        .eq("id", property_id)
        .eq("user_id", user_id)
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=404, detail="Property not found")

File: lib/properties.ts
import "server-only";

import { createServerSupabaseClient } from "@/lib/supabase-server";
import { Property, PropertyFilters, PropertyFormData } from "@/types";

const TABLE = "properties";

type PropertyPayload = Partial<
  Omit<PropertyFormData, "image_url" | "notes" | "listing_url">
> & {
  image_url?: string | null;
  notes?: string | null;
  listing_url?: string | null;
};

function normalizeSearch(search: string) {
  return search.replace(/[%,()]/g, " ").trim();
}

function cleanPayload(data: Partial<PropertyFormData>): PropertyPayload {
  const payload: PropertyPayload = { ...data };

  for (const key of ["image_url", "notes", "listing_url"] as const) {
    if (payload[key] === "") {
      payload[key] = null;
    }
  }

  return payload;
}

export async function getProperties(
  filters: PropertyFilters = {}
): Promise<Property[]> {
  const supabase = createServerSupabaseClient();
  let query = supabase.from(TABLE).select("*");

  if (filters.search) {
    const search = normalizeSearch(filters.search);
    if (search) {
      query = query.or(`address.ilike.%${search}%,suburb.ilike.%${search}%`);
    }
  }

  if (filters.bedrooms) {
    const bedrooms = Number(filters.bedrooms);
    if (!Number.isNaN(bedrooms)) {
      query =
        bedrooms >= 3
          ? query.gte("bedrooms", bedrooms)
          : query.eq("bedrooms", bedrooms);
    }
  }

  if (filters.max_price) {
    const maxPrice = Number(filters.max_price);
    if (!Number.isNaN(maxPrice)) {
      query = query.lte("price_per_week", maxPrice);
    }
  }

  if (filters.status) {
    query = query.eq("status", filters.status);
  }

  const { data, error } = await query.order("created_at", {
    ascending: false,
  });

  if (error) {
    throw new Error(error.message);
  }

  return data ?? [];
}

export async function getProperty(id: string): Promise<Property> {
  const supabase = createServerSupabaseClient();
  const { data, error } = await supabase
    .from(TABLE)
    .select("*")
    .eq("id", id)
    .maybeSingle();

  if (error) {
    throw new Error(error.message);
  }

  if (!data) {
    throw new Error("Property not found");
  }

  return data;
}

export async function createProperty(data: PropertyFormData): Promise<Property> {
  const supabase = createServerSupabaseClient();
  const { data: property, error } = await supabase
    .from(TABLE)
    .insert(cleanPayload(data))
    .select("*")
    .single();

  if (error) {
    throw new Error(error.message);
  }

  return property;
}

export async function updateProperty(
  id: string,
  data: Partial<PropertyFormData>
): Promise<Property> {
  const supabase = createServerSupabaseClient();
  const { data: property, error } = await supabase
    .from(TABLE)
    .update(cleanPayload(data))
    .eq("id", id)
    .select("*")
    .maybeSingle();

  if (error) {
    throw new Error(error.message);
  }

  if (!property) {
    throw new Error("Property not found");
  }

  return property;
}

export async function deleteProperty(id: string): Promise<void> {
  const supabase = createServerSupabaseClient();
  const { data, error } = await supabase
    .from(TABLE)
    .delete()
    .eq("id", id)
    .select("id")
    .maybeSingle();

  if (error) {
    throw new Error(error.message);
  }

  if (!data) {
    throw new Error("Property not found");
  }
}
